core/external_search.py

```diff
@@ -3627,16 +2784,49 @@ def __init__(self, *args, **kwargs):
         super().__init__(*args, **kwargs)
         self.search_index_client = search_index_client or ExternalSearchIndex(self._db)
 
-    def process_batch(self, works):
-        """
-        :return: a mixed list of Works and CoverageFailure objects.
-        """
-        successes, failures = self.search_index_client.bulk_update(works)
+        #
+        # Try to migrate to the latest schema. If the function returns None, it means
+        # that no migration is necessary, and we're already at the latest version. If
+        # we're already at the latest version, then simply upload search documents instead.
+        #
+        self.receiver = None
+        self.migration: Optional[
+            SearchMigrationInProgress
+        ] = self.search_index_client.start_migration()
+        if self.migration is None:
+            self.receiver: SearchDocumentReceiver = (
+                self.search_index_client.start_updating_search_documents()
+            )
+        else:
+            # We do have a migration, we must clear out the index and repopulate the index
+            self.remove_search_coverage_records()
+
+    def on_completely_finished(self):
+        super().on_completely_finished()
+        # Tell the search migrator that no more documents are going to show up.
+        target: SearchDocumentReceiverType = self.migration or self.receiver
+        target.finish()
```

Comment:
It doesn't look like this `on_completely_finished` will get called, because `Scripts.run`, rather than `BaseCoverageProvider.run`, is the runner here.

Reply:
I just assumed this was in working order :sweat_smile: 
I've updated the code, and written a test for it.

Reply:
This one is on me. The scripts and coverage providers are a complete nightmare of subclassing with random implementation methods being overridden. I couldn't tell what was calling what...

Reply:
@RishiDiwanTT Your changes on [Fixed the migration runner script](https://github.com/ThePalaceProject/circulation/pull/1268/commits/2984266147247972ff8d265c6f96ffd0869736c1) look like they resolve this problem.